src/zjs_aio_k64f.c

```diff
@@ -0,0 +1,331 @@
+    zjs_obj_add_number(pinobj, "pin", pin);
+
+    aio_handle_t *handle = aio_alloc_handle();
+    if (!handle) {
+        return zjs_error("could not allocate handle");
+    }
+
+    // lookup what device the pins are connected to
+    // A0 and A1 are connected to ADC_0, A2 and A3 are connected to ADC_1
+    struct device *dev = adc_device[map_device(pin)];
+    if (!dev) {
+        const char *dev_name = map_device_name(pin);
+        if (dev_name) {
+            dev = device_get_binding(dev_name);
+            if (!dev) {
+                zjs_free(handle);
+                return zjs_error("failed to initialize AIO device");
+            }
+            adc_enable(dev);
+            adc_device[map_device(pin)] = dev;
+        } else {
+            zjs_free(handle);
+            return zjs_error("failed to find ADC device");
+        }
+    }
+
+    pin_enabled[pin - AIO_MIN]++;
+    DBG_PRINT("AIO pin %d refcount %d\n", pin, pin_enabled[pin-12]);
+
+    handle->pin_obj = jerry_acquire_value(pinobj);
+    handle->pin = pin;
+
+    // make it an emitter object
+    zjs_make_emitter(pinobj, zjs_aio_prototype, handle, aio_free_cb);
+
+    // add to the list of opened handles
+    ZJS_LIST_APPEND(aio_handle_t, opened_handles, handle);
+
+    return pinobj;
+}
+
+static s32_t aio_poll_routine(void *h)
+{
+    u32_t uptime = k_uptime_get_32();
+
+    for (int i = 0; i < AIO_LEN; i++) {
+        if (pin_enabled[i]) {
+            // FIXME: We dont know when a object subscribes to onchange
+            // events, need a way to intercept the .on()
+            // currently, if the pin is opened, then it do a read
+            // and signal the event, this could slow down the system
+            // when there are multiple AIO pins opened.
+            if (uptime % (u32_t)(CONFIG_SYS_CLOCK_TICKS_PER_SEC /
+                AIO_POLL_FREQUENCY * 10) == 0) {
+                pin_values[i] = pin_read(i + AIO_MIN);
+
+                if (pin_values[i] != pin_last_values[i]) {
+                    // send updates only if value has changed
+                    aio_handle_t *h = ZJS_LIST_FIND(aio_handle_t,
+                                                    opened_handles, pin,
+                                                    i + AIO_MIN);
+                    if (h) {
+                        ZVAL val = jerry_create_number(pin_values[i]);
+                        zjs_defer_emit_event(h->pin_obj, "change",
+                                             &val,
+                                             sizeof(val),
+                                             zjs_copy_arg,
+                                             zjs_release_args);
+                    }
+                    pin_last_values[i] = pin_values[i];
+                }
+            }
+        }
+    }
+
+    zjs_loop_unblock();
+    return K_FOREVER;
+}
+
+jerry_value_t zjs_aio_init()
+{
+    // FIXME: need a way to unregister service routines
```

Comment:
True, Brian should care about this for ashell; could you open an issue?

Reply:
I'll open an issue